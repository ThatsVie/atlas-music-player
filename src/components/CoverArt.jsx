import React from 'react';
import RetroPug from '../assets/RetroPug.png';

const CoverArt = () => {
  return (
    <div className="relative mb-6 aspect-square">
      <img
        src={RetroPug}
        alt="A vibrant 90s-inspired pug wearing a colorful jacket, sunglasses, and a cap with a geometric-patterned background."
        className="h-full w-full rounded-md object-cover block align-middle"
      />
    </div>
  );
};

export default CoverArt;
