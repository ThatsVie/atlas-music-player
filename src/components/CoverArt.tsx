import React, { useState, useEffect, useRef } from 'react';

interface CoverArtProps {
  cover: string;
  lyrics?: string | null;
}

const CoverArt: React.FC<CoverArtProps> = ({ cover, lyrics }) => {
  const [isHovered, setIsHovered] = useState(false);
  const lyricsRef = useRef<HTMLDivElement>(null);

  useEffect(() => {
    if (lyricsRef.current) {
      console.log('Resetting scroll position for lyrics container.');
      lyricsRef.current.scrollTop = 0;
      console.log(
        'Lyrics container scroll position after reset:',
        lyricsRef.current.scrollTop
      );
    }
  }, [lyrics]);

  return (
    <div
      className='relative mb-6 aspect-square'
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
    >
      <img
        src={cover}
        alt='Album cover'
        className='h-full w-full rounded-md object-cover block align-middle'
      />

      {isHovered && lyrics && (
        <div
          ref={lyricsRef}
          className='absolute inset-0 bg-black bg-opacity-70 text-white p-4 rounded-md flex flex-col justify-start overflow-auto z-10'
          style={{
            maxHeight: '100%',
            boxSizing: 'border-box',
          }}
        >
          <p className='text-sm leading-6 whitespace-pre-line p-0 m-0 box-border'>
            {lyrics}
          </p>
        </div>
      )}
    </div>
  );
};

export default CoverArt;
