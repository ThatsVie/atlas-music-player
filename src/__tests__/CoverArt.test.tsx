import { render, fireEvent, screen } from '@testing-library/react';
import { describe, test, expect } from 'vitest';
import CoverArt from '@/components/CoverArt';

describe('CoverArt Component', () => {
  test('renders correctly with only cover image', () => {
    // Arrange
    const coverImage = 'https://example.com/cover.jpg';

    // Act
    const { container } = render(<CoverArt cover={coverImage} />);

    // Assert
    expect(container).toMatchSnapshot();
  });

  test('renders correctly with cover image and lyrics', () => {
    // Arrange
    const coverImage = 'https://example.com/cover.jpg';
    const lyrics = 'Sample lyrics';

    // Act
    const { container } = render(
      <CoverArt cover={coverImage} lyrics={lyrics} />
    );

    // Assert
    expect(container).toMatchSnapshot();
  });

  test('resets lyrics scroll position when lyrics change', async () => {
    // Arrange
    const coverImage = 'https://example.com/cover.jpg';
    const { rerender } = render(<CoverArt cover={coverImage} lyrics="Initial lyrics" />);

    // Act
    const coverContainer = screen.getByRole('img', { name: /album cover/i }).parentElement!;
    fireEvent.mouseEnter(coverContainer);

    const lyricsDiv = await screen.findByText(/Initial lyrics/i);
    const lyricsContainer = lyricsDiv.parentElement as HTMLDivElement;

    lyricsContainer.scrollTop = 50;

    // Act
    rerender(<CoverArt cover={coverImage} lyrics="Updated lyrics" />);

    // Assert
    expect(lyricsContainer.scrollTop).toBe(0);
  });

  test('shows lyrics overlay on hover when lyrics exist', async () => {
    // Arrange
    const coverImage = 'https://example.com/cover.jpg';
    const lyrics = 'Sample lyrics';

    // Act
    render(<CoverArt cover={coverImage} lyrics={lyrics} />);

    const coverContainer = screen.getByRole('img', { name: /album cover/i }).parentElement!;
    fireEvent.mouseEnter(coverContainer);

    // Assert
    expect(await screen.findByText(/Sample lyrics/i)).toBeInTheDocument();
  });

  test('hides lyrics overlay when mouse leaves', async () => {
    // Arrange
    const coverImage = 'https://example.com/cover.jpg';
    const lyrics = 'Sample lyrics';

    // Act
    render(<CoverArt cover={coverImage} lyrics={lyrics} />);

    const coverContainer = screen.getByRole('img', { name: /album cover/i }).parentElement!;
    fireEvent.mouseEnter(coverContainer);

    expect(await screen.findByText(/Sample lyrics/i)).toBeInTheDocument();

    // Act
    fireEvent.mouseLeave(coverContainer);

    // Assert
    expect(screen.queryByText(/Sample lyrics/i)).not.toBeInTheDocument();
  });

  test('does not show lyrics overlay on hover when no lyrics are provided', async () => {
    // Arrange
    const coverImage = 'https://example.com/cover.jpg';

    // Act
    render(<CoverArt cover={coverImage} lyrics={null} />);

    const coverContainer = screen.getByRole('img', { name: /album cover/i }).parentElement!;
    fireEvent.mouseEnter(coverContainer);

    // Assert
    expect(screen.queryByText(/Sample lyrics/i)).not.toBeInTheDocument();
  });
});
