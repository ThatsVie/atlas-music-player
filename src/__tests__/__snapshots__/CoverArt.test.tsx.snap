// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CoverArt Component > renders correctly with cover image and lyrics 1`] = `
<div>
  <div
    class="relative mb-6 aspect-square"
  >
    <img
      alt="Album cover"
      class="h-full w-full rounded-md object-cover block align-middle"
      src="https://example.com/cover.jpg"
    />
  </div>
</div>
`;

exports[`CoverArt Component > renders correctly with only cover image 1`] = `
<div>
  <div
    class="relative mb-6 aspect-square"
  >
    <img
      alt="Album cover"
      class="h-full w-full rounded-md object-cover block align-middle"
      src="https://example.com/cover.jpg"
    />
  </div>
</div>
`;
